contracts/yellow/VaultImpl.sol

```diff
@@ -1,11 +1,267 @@
+     * @param payload Encoded payload, which consists of rid (unique identifier id), expire timestamp, destination address and an array of allocations.
+     * @param brokerSignature Payload signed by the Broker.
+     * @return bool Return 'true' if withdrawn successfully.
+     */
+    function _withdraw_interactions(
+        address account,
+        bytes memory payload,
+        bytes memory brokerSignature
+    ) internal returns (bool) {
+        bytes32 sigHash = keccak256(brokerSignature);
+        (bytes32 rid, , address destination, Allocation[] memory assets) = _extractPayload(account, sigHash, payload);
+
+        _sigUsage[account][sigHash] = true;
+
+        for (uint256 i = 0; i < assets.length; i++) {
+            _transferAssetTo(assets[i].asset, destination, assets[i].amount);
+            _ledgerId.increment();
+            emit Withdrawn(_ledgerId.current(), destination, assets[i].asset, assets[i].amount, rid);
+        }
+
+        return true;
+    }
+
+    /**
+     * @notice Check supplied signatures to be indeed signed by Broker and OTP service.
+     * @param action Action type. One of DEPOSIT_TYPE and WITHDRAW_TYPE.
+     * @param payload Encoded payload, which consists of rid (unique identifier id), expire timestamp, destination address and an array of allocations.
+     * @param brokerSignature Broker signature.
+     * @param otpSignature OTP signature.
+     */
+    function _requireValidInput(
+        bytes32 action,
+        bytes memory payload,
+        bytes memory brokerSignature,
+        bytes memory otpSignature
+    ) internal view {
+        require(action != 0, 'Vault: action is required');
+
+        bytes32 digest = ECDSA.toEthSignedMessageHash(
+            keccak256(abi.encodePacked(action, payload))
+        );
+        address recoveredBrokerAddress = ECDSA.recover(digest, brokerSignature);
+        require(recoveredBrokerAddress == _brokerKeyDerivedAddress, 'Vault: invalid broker signature');
+        address recoveredOTPAddress = ECDSA.recover(digest, otpSignature);
+        require(recoveredOTPAddress == _otpKeyDerivedAddress, 'Vault: invalid OTP signature');
+    }
+
+    /**
+     * @notice Internal function to extract payload data.
+     * @param account Account address.
+     * @param sigHash Broker signature keccak256 hash.
+     * @param payload Encoded payload, which consists of rid (unique identifier id), expire timestamp, destination address and an array of allocations.
+     * @return bytes32 rid (unique identifier id).
+     * @return uint64 expire timestamp.
+     * @return address destination address.
+     * @return Asset Array of allocations.
+     */
+    function _extractPayload(
+        address account,
+        bytes32 sigHash,
+        bytes memory payload
+    ) internal view returns (bytes32, uint64, address, Allocation[] memory) {
+        (bytes32 rid, uint64 expire, address destination, Allocation[] memory assets) = abi.decode(
```

Comment:
- expire can be type uint48.
- need to add the chainId, to ensure we can't re-use a payload on an other chain.

Reply:
Will also add `implAddress` to disable breaking stuff when a request with `expire` has been issued, but not yes used, while Impl was updated.

Reply:
68d0b0199c3213fd4f6726ee4543adc8b162d0a6

Reply:
@calj we cannot natively support uint48 in Golang

Reply:
24da6a7486d8bdcd60885e40c73d8dcd12dc2f88